File: web/src/lib/billmath.test.js
// parity tests -- same numbers as android BillMathTest.kt (real sample bill: net 24349 etc)
import { describe, it, expect } from 'vitest'
import { leg, ta, makeTrip, billTotals, legDefaults, amountInWords } from './billmath.js'

// real sample: decoded official bill xlsx batches two finished trips --
// trip1: 11 May same-day Dhaka (legs 441+482), claimed 0 nights/0 food ("-" on the sheet)
// trip2: 8-14 Jun Bhola (legs 176+1500), 6 nights, 6.5 food days (A** category span)
it('real sample bill with claimed overrides', () => {
  const trip1 = makeTrip([leg(441), leg(482)], '2026-05-11', '2026-05-11', 0, 0)
  const trip2 = makeTrip([leg(176), leg(1500)], '2026-06-08', '2026-06-14', 6, 6.5)
  const totals = billTotals([trip1, trip2])
  expect(totals.ta).toBeCloseTo(2599, 3)
  expect(totals.accommodation).toBeCloseTo(12000, 3)
  expect(totals.food).toBeCloseTo(9750, 3)
  expect(totals.net).toBeCloseTo(24349, 3)
  expect(amountInWords(Math.round(totals.net))).toBe('Twenty Four Thousand Three Hundred Forty Nine')
})

it('unresolved trip defaults to zero', () => {
  const trip1 = makeTrip([leg(441)], '2026-05-11', '2026-05-11')
  expect(trip1.resolvedNights).toBe(0)
  expect(trip1.resolvedFood).toBeCloseTo(0, 4)
})

it('ta sums leg fares', () => {
  expect(ta([leg(176), leg(1500), leg(441)])).toBeCloseTo(2117, 3)
})

it('billTotals sums resolved nights/food across trips', () => {
  const totals = billTotals([makeTrip([leg(500)], '2026-06-01', '2026-06-01', 0, 0.5)])
  expect(totals.accommodation).toBeCloseTo(0, 3)
  expect(totals.food).toBeCloseTo(750, 3)
  expect(totals.ta).toBeCloseTo(500, 3)
})

it('legDefaults day-grouped, last day halved', () => {
  const legs = [
    leg(100, '2026-06-08'), // day 1, first leg -> night
    leg(50, '2026-06-08'), // day 1, second leg -> no double count
    leg(100, '2026-06-14'), // last day -> half food, no night
  ]
  expect(legDefaults(legs, '2026-06-14')).toEqual([[1, 1], [0, 0], [0, 0.5]])
})

it('legDefaults single day trip', () => {
  expect(legDefaults([leg(100, '2026-06-01')], '2026-06-01')).toEqual([[0, 0.5]])
})

describe('amountInWords', () => {
  it('zero', () => expect(amountInWords(0)).toBe('Zero'))
  it('one lakh', () => expect(amountInWords(100000)).toBe('One Lakh'))
  it('one lakh five', () => expect(amountInWords(100005)).toBe('One Lakh Five'))
  it('crore', () => expect(amountInWords(12345678)).toBe(
    'One Crore Twenty Three Lakh Forty Five Thousand Six Hundred Seventy Eight'))
})

File: android/app/src/test/java/bd/sicip/qavisit/domain/BillMathTest.kt
// TA/DA bill math: night/food span totals, TA sum, amount-in-words
package bd.sicip.qavisit.domain

import org.junit.Assert.assertEquals
import org.junit.Test

class BillMathTest {

    // real sample: tada-template.xlsx ToT Monitoring visit, 08-14 Jun, 6 nights
    @Test fun real_sample_trip_span_totals() {
        val legs = listOf(Leg(fare = 176.0), Leg(fare = 1500.0))
        val totals = billTotals(legs, "2026-06-08", "2026-06-14")
        assertEquals(6, tripNights("2026-06-08", "2026-06-14"))
        assertEquals(6.5, tripFoodDays("2026-06-08", "2026-06-14"), 0.0001)
        assertEquals(12000.0, totals.accommodation, 0.001)
        assertEquals(9750.0, totals.food, 0.001)
        assertEquals(1676.0, totals.ta, 0.001)
        assertEquals(1676.0 + 12000.0 + 9750.0, totals.net, 0.001)
    }

    // TA = sum of fares, sample values lifted from the real template
    @Test fun ta_sums_leg_fares() {
        val legs = listOf(Leg(fare = 176.0), Leg(fare = 1500.0), Leg(fare = 441.0))
        assertEquals(2117.0, ta(legs), 0.001)
    }

    @Test fun single_day_trip_no_night_half_food() {
        assertEquals(0, tripNights("2026-06-01", "2026-06-01"))
        assertEquals(0.5, tripFoodDays("2026-06-01", "2026-06-01"), 0.0001)
        val totals = billTotals(listOf(Leg(fare = 500.0)), "2026-06-01", "2026-06-01")
        assertEquals(0.0, totals.accommodation, 0.001)
        assertEquals(750.0, totals.food, 0.001)
        assertEquals(500.0, totals.ta, 0.001)
    }

    // per-leg defaults: day-grouped, first leg of a day carries the value, last day halved
    @Test fun legDefaults_day_grouped_last_day_halved() {
        val legs = listOf(
            Leg(fare = 100.0, depDate = "2026-06-08"), // day 1, first leg -> night
            Leg(fare = 50.0, depDate = "2026-06-08"),  // day 1, second leg -> no double count
            Leg(fare = 100.0, depDate = "2026-06-14"), // last day -> half food, no night
        )
        val defaults = legDefaults(legs, "2026-06-08", "2026-06-14")
        assertEquals(listOf(1 to 1.0, 0 to 0.0, 0 to 0.5), defaults)
    }

    @Test fun legDefaults_single_day_trip() {
        val legs = listOf(Leg(fare = 100.0, depDate = "2026-06-01"))
        assertEquals(listOf(0 to 0.5), legDefaults(legs, "2026-06-01", "2026-06-01"))
    }

    // ---- amountInWords ----
    @Test fun amountInWords_real_sample() {
        assertEquals("Twenty Four Thousand Three Hundred Forty Nine", amountInWords(24349))
    }

    @Test fun amountInWords_zero() {
        assertEquals("Zero", amountInWords(0))
    }

    @Test fun amountInWords_one_lakh() {
        assertEquals("One Lakh", amountInWords(100000))
    }

    @Test fun amountInWords_crore() {
        assertEquals(
            "One Crore Twenty Three Lakh Forty Five Thousand Six Hundred Seventy Eight",
            amountInWords(12345678)
        )
    }
}
